consensus/src/runway/collection.rs

```diff
@@ -0,0 +1,536 @@
+                write!(f, "mismatched salt, expected {}, got {}", expected, got)
+            }
+            InvalidUnit(e) => write!(f, "invalid unit: {}", e),
+            ForeignUnit(id) => write!(f, "unit from node {:?}", id),
+        }
+    }
+}
+
+impl<H: Hasher, D: Data, S: Signature> From<ValidationError<H, D, S>> for Error<H, D, S> {
+    fn from(ve: ValidationError<H, D, S>) -> Self {
+        Error::InvalidUnit(ve)
+    }
+}
+
+impl<H: Hasher, D: Data, S: Signature> From<SignatureError<NewestUnitResponse<H, D, S>, S>>
+    for Error<H, D, S>
+{
+    fn from(_: SignatureError<NewestUnitResponse<H, D, S>, S>) -> Self {
+        Error::WrongSignature
+    }
+}
+
+/// The status of an ongoing collection.
+#[derive(PartialEq, Debug)]
+pub enum Status {
+    /// Received less than threshold responses.
+    Pending,
+    /// Received at least threshold responses.
+    Ready(Round),
+    /// Received all possible responses.
+    Finished(Round),
+}
+
+/// Initial unit collection to figure out at which round we should start unit production.
+/// Unfortunately this isn't quite BFT, but it's good enough in many situations.
+pub struct Collection<'a, MK: KeyBox> {
+    keychain: &'a MK,
+    validator: &'a Validator<'a, MK>,
+    starting_round: Round,
+    responders: HashSet<NodeIndex>,
+    threshold: NodeCount,
+    salt: Salt,
+}
+
+impl<'a, MK: KeyBox> Collection<'a, MK> {
+    /// Create a new collection instance ready to collect responses.
+    /// The returned salt should be used to initiate newest unit requests.
+    pub fn new(
+        keychain: &'a MK,
+        validator: &'a Validator<'a, MK>,
+        threshold: NodeCount,
+    ) -> (Self, Salt) {
+        let salt = generate_salt();
+        let responders = [keychain.index()].into();
+        (
+            Collection {
+                keychain,
+                validator,
+                starting_round: 0,
+                responders,
+                threshold,
+                salt,
+            },
+            salt,
+        )
+    }
+
+    /// Process a response to a newest unit request.
+    pub fn on_newest_response<H: Hasher, D: Data>(
+        &mut self,
+        unchecked_response: UncheckedSigned<NewestUnitResponse<H, D, MK::Signature>, MK::Signature>,
+    ) -> Result<Status, Error<H, D, MK::Signature>> {
+        let response = unchecked_response.check(self.keychain)?.into_signable();
+        if response.salt != self.salt {
+            return Err(Error::SaltMismatch(self.salt, response.salt));
+        }
+        if let Some(unchecked_unit) = response.unit {
+            let checked_signed_unit = self.validator.validate_unit(unchecked_unit)?;
+            let checked_unit = checked_signed_unit.as_signable();
+            if checked_unit.creator() != self.keychain.index() {
+                return Err(Error::ForeignUnit(checked_unit.creator()));
+            }
+            let starting_round_candidate = checked_unit.round() + 1;
```

Comment:
Nit: Could be 
```
self.starting_round = std::cmp::max(self.starting_round, checked_unit.round() + 1);
```
but I'm not sure if it is popular to use max in rust.

Reply:
Eh, I dunno, I think we were trying to avoid that to be more explicit, although maybe this is a strange leftover from the Go days...

Reply:
Do you think this would be more readable?

Reply:
I usually prefer using `max` as it is more readable to me. But I'm completely fine with current approach it's a very small nit and I'm not sure how readable it is for others.

Reply:
Added then.